### PR TITLE
BUG: fix col iteration in DataFrame.round, #11611

### LINKED ISSUE #11611: Stack overflow on applying numpy functions to DataFrame with duplicated column entries.
Applying a numpy function, such as `np.round`, to a DataFrame with duplicated column indicies can cause an unrecoverable stack overflow error (Fatal Python error: Cannot recover from stack overflow.), which causes e.g. a ipython kernel to crash. E.g. take the following example, where python crashes at line 4:

```python
x = pd.DataFrame(np.random.randn(3,3))
y = pd.DataFrame(np.random.randn(3,3))
z = pd.concat((x, y), axis=1)
print(np.round(z))
```

However, removing the duplicate column entries, works as expected:
```python
...
z = pd.concat((x, y), axis=1, ignore_index=True)
print(np.round(z))
```

---
python 3.5.0, numpy 1.10.1, pandas 0.17.0

### PR DESCRIPTION
**Pull request by skycaptain**
Fixes #11611. The iterator now uses DataFrame.iteritems instead of direct indexing.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4382,17 +4382,17 @@ def round(self, decimals=0, out=None):
         from pandas.tools.merge import concat
 
         def _dict_round(df, decimals):
-            for col in df:
+            for col, vals in df.iteritems():
                 try:
-                    yield np.round(df[col], decimals[col])
+                    yield np.round(vals, decimals[col])
```
Comment (maximilianr) [inline pandas/core/frame.py]: Is the decimals lookup going to have the same problem if it's not unique here?

**Comment by jreback** [comment]
need some tests

File: pandas/core/frame.py (hunk 1 above)
Comment (skycaptain) [inline pandas/core/frame.py]: The value of decimals is either of type dict or pandas.Series. Whereas the former can not have duplicate entries, a pandas.Series might have and direct indexing will return all values with the same index. So, this case is still not handled.

Comment (jreback) [inline pandas/core/frame.py]: I think let's raise if ``decimals`` is not unique (it should be a dict/Series).

Comment (skycaptain) [inline pandas/core/frame.py]: Had the same thought this morning.

**Comment by jreback** [comment]
pls add the test example as well (you can just smoke tests it I think; e.g. assert it runs).

add a tests for passing a non-unique decimal.

put this in the bug-fix section for a whatsnew entry.

**Comment by skycaptain** [comment]
Agreed.

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/80a2d530bae32ef7ecf93aa275d17659ee8e8ac1

thanks!

**Pull request closed without merging**